File: src/engine/muxer.ts
import os from "node:os";
import ms from "ms";
import { execa } from "execa";
import path from "path";
import fs from "fs-extra";
import { Logger } from "../core/logger.js";
import { getErrorMessage } from "../types/errors.js";

/**
 * Trim options for timestamp-based segment selection.
 * These apply FFmpeg -ss/-to flags during mux to trim segment boundaries.
 */
export interface MuxTrimOptions {
  /** Offset into the first segment where the actual start time falls (seconds) */
  trimStartOffset?: number;
  /** Duration of the output from the trim start (seconds) */
  trimDuration?: number;
  /** Video bitrate for precise re-encode (kbps) — used for ABR modes */
  videoBitrate?: number;
  /** Audio bitrate for precise re-encode (kbps) */
  audioBitrate?: number;
  /** CRF value for quality-based encoding (0-51, lower = better, 18 = perceptually lossless) */
  crf?: number;
  /** Use precise trim (re-encode for exact duration) - default true */
  usePreciseTrim?: boolean;
  /** Use 2-pass encoding (only for ABR mode, not CRF) */
  twoPass?: boolean;
}

export class Muxer {
  static async mux(
    videoPath: string,
    audioPath: string | null,
    outputPath: string,
    signal?: AbortSignal,
    trimOptions?: MuxTrimOptions,
  ): Promise<void> {
    const logger = Logger.getInstance();

    if (signal?.aborted) {
      throw new DOMException("Aborted", "AbortError");
    }

    // Normalize paths for the current platform
    const normalizedVideoPath = path.resolve(videoPath);
    const normalizedAudioPath = audioPath ? path.resolve(audioPath) : null;
    const normalizedOutputPath = path.resolve(outputPath);

    // Verify input files exist
    if (!(await fs.pathExists(normalizedVideoPath))) {
      throw new Error(`Video file not found: ${normalizedVideoPath}`);
    }
    if (normalizedAudioPath && !(await fs.pathExists(normalizedAudioPath))) {
      throw new Error(`Audio file not found: ${normalizedAudioPath}`);
    }

    // Ensure output directory exists
    await fs.ensureDir(path.dirname(normalizedOutputPath));

    // Decide: codec copy (fast but imprecise) vs re-encode (precise)
    const hasTrim = trimOptions?.trimStartOffset || trimOptions?.trimDuration;
    const usePrecise =
      trimOptions?.usePreciseTrim !== false && hasTrim &&
      (trimOptions?.crf != null || trimOptions?.videoBitrate || trimOptions?.audioBitrate);

    if (usePrecise && trimOptions?.twoPass && trimOptions.videoBitrate) {
      // 2-pass ABR re-encode (for target bitrate with optimal distribution)
      await this.runTwoPassEncode(
        normalizedVideoPath,
        normalizedAudioPath,
        normalizedOutputPath,
        trimOptions,
        signal,
      );
    } else {
      // Build FFmpeg arguments for single-pass paths
      const ffmpegArgs = this.buildInputArgs(
        normalizedVideoPath,
        normalizedAudioPath,
        trimOptions ?? {},
      );

      if (usePrecise) {
        // Single-pass re-encode for exact duration
        this.appendEncodeArgs(ffmpegArgs, normalizedAudioPath, trimOptions!);
        ffmpegArgs.push("-movflags", "faststart", normalizedOutputPath);
      } else {
        // Standard codec copy (fast)
        ffmpegArgs.push(
          "-c",
          "copy",
          "-movflags",
          "faststart",
          normalizedOutputPath,
        );
      }

      await this.runFFmpeg(ffmpegArgs, signal);
    }
  }

  /**
   * Run a 2-pass x264 encode for better bitrate distribution
   */
  private static async runTwoPassEncode(
    videoPath: string,
    audioPath: string | null,
    outputPath: string,
    trimOptions: MuxTrimOptions,
    signal?: AbortSignal,
  ): Promise<void> {
    const logger = Logger.getInstance();
    const passlogPrefix = path.join(
      os.tmpdir(),
      `moombox-2pass-${Date.now()}`,
    );

    try {
      // === Pass 1: Analysis (no audio, output to null) ===
      logger.info("[Muxer] 2-pass encode: starting pass 1 (analysis)");

      const pass1Args = this.buildInputArgs(videoPath, audioPath, trimOptions);
      pass1Args.push(
        "-c:v", "libx264",
        "-b:v", `${trimOptions.videoBitrate}k`,
        "-preset", "fast",
        "-pass", "1",
        "-passlogfile", passlogPrefix,
        "-an",
        "-f", "null",
        os.devNull,
      );

      await this.runFFmpeg(pass1Args, signal);

      // Check abort between passes
      if (signal?.aborted) {
        throw new DOMException("Aborted", "AbortError");
      }

      // === Pass 2: Encode with optimized bitrate distribution ===
      logger.info("[Muxer] 2-pass encode: starting pass 2 (encode)");

      const pass2Args = this.buildInputArgs(videoPath, audioPath, trimOptions);
      pass2Args.push(
        "-c:v", "libx264",
        "-b:v", `${trimOptions.videoBitrate}k`,
        "-preset", "fast",
        "-pass", "2",
        "-passlogfile", passlogPrefix,
      );

      // Audio encoding (same logic as single-pass)
      if (audioPath && trimOptions.audioBitrate) {
        pass2Args.push("-c:a", "aac", "-b:a", `${trimOptions.audioBitrate}k`);
      } else if (audioPath) {
        pass2Args.push("-c:a", "copy");
      }

      pass2Args.push("-movflags", "faststart", outputPath);

      await this.runFFmpeg(pass2Args, signal);

      logger.info("[Muxer] 2-pass encode: complete");
    } finally {
      // Clean up passlog files
      for (const suffix of ["-0.log", "-0.log.mbtree"]) {
        try {
          await fs.remove(passlogPrefix + suffix);
        } catch {
          // Ignore cleanup errors
        }
      }
    }
  }

  /**
   * Build common input arguments (seek, inputs, duration)
   */
  private static buildInputArgs(
    videoPath: string,
    audioPath: string | null,
    trimOptions: MuxTrimOptions,
  ): string[] {
    const args = ["-y"];

    if (trimOptions.trimStartOffset && trimOptions.trimStartOffset > 0) {
      args.push("-ss", String(trimOptions.trimStartOffset));
    }

    args.push("-i", videoPath);

    if (audioPath) {
      if (trimOptions.trimStartOffset && trimOptions.trimStartOffset > 0) {
        args.push("-ss", String(trimOptions.trimStartOffset));
      }
      args.push("-i", audioPath);
    }

    if (trimOptions.trimDuration && trimOptions.trimDuration > 0) {
      args.push("-t", String(trimOptions.trimDuration));
    }

    return args;
  }

  /**
   * Append single-pass video/audio encode arguments
   */
  private static appendEncodeArgs(
    args: string[],
    audioPath: string | null,
    trimOptions: MuxTrimOptions,
  ): void {
    const logger = Logger.getInstance();

    if (trimOptions.crf != null) {
      // CRF mode: quality-based encoding — x264 decides bitrate per-frame
      logger.info(`[Muxer] Using CRF ${trimOptions.crf} encode (quality-based, slow preset)`);
      args.push(
        "-c:v", "libx264",
        "-crf", String(trimOptions.crf),
        "-preset", "slow",
      );
      // Re-encode audio at source bitrate to keep sync with re-encoded video
      if (trimOptions.audioBitrate) {
        args.push("-c:a", "aac", "-b:a", `${trimOptions.audioBitrate}k`);
      } else {
        args.push("-c:a", "aac");
      }
    } else {
      // ABR mode: target bitrate encoding
      logger.info("[Muxer] Using precise trim (re-encode for exact duration)");

      if (trimOptions.videoBitrate) {
        args.push(
          "-c:v", "libx264",
          "-b:v", `${trimOptions.videoBitrate}k`,
          "-preset", "fast",
        );
      } else {
        args.push("-c:v", "copy");
      }

      if (audioPath && trimOptions.audioBitrate) {
        args.push("-c:a", "aac", "-b:a", `${trimOptions.audioBitrate}k`);
      } else if (audioPath) {
        args.push("-c:a", "copy");
      }
    }
  }

  /**
   * Execute FFmpeg with standard error handling
   */
  private static async runFFmpeg(
    args: string[],
    signal?: AbortSignal,
  ): Promise<void> {
    const logger = Logger.getInstance();
    logger.debug(
      `[Muxer] Running: ffmpeg ${args.map((a) => `"${a}"`).join(" ")}`,
    );

    try {
      await execa("ffmpeg", args, {
        stdin: "ignore",
        cancelSignal: signal,
        timeout: ms("10m"),
        cleanup: true,
      });

      logger.debug(`[Muxer] FFmpeg completed successfully`);
    } catch (error: unknown) {
      const execaErr = error as { isCanceled?: boolean; name?: string; stderr?: string; exitCode?: number; message?: string };
      if (execaErr.isCanceled || execaErr.name === 'AbortError') {
        throw new DOMException("Aborted", "AbortError");
      }

      const stderr = execaErr.stderr || getErrorMessage(error);
      logger.error(`[Muxer] FFmpeg failed: ${stderr.slice(-500)}`);
      throw new Error(`FFmpeg exited with code ${execaErr.exitCode || 'unknown'}: ${stderr.slice(-200)}`);
    }
  }
}

File: internal/engine/muxer.go
package engine

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
)

// TrimOptions configures trimming and encoding during muxing.
type TrimOptions struct {
	TrimStartOffset float64               // Seconds into first segment
	TrimDuration    float64               // Total duration from trim start
	VideoBitrate    int                   // kbps for ABR mode (0 = no target)
	AudioBitrate    int                   // kbps
	CRF             int                   // 0-51, 18 = perceptually lossless (0 = copy)
	UsePreciseTrim  bool                  // Default true
	TwoPass         bool                  // Only for ABR with VideoBitrate
	ProgressFn      func(percent float64) // Optional: called with 0-100 as FFmpeg encodes
}

// Muxer handles FFmpeg operations.
type Muxer struct {
	ffmpegPath  string
	ffprobePath string
	logger      interface {
		Debug(msg string, args ...any)
		Info(msg string, args ...any)
		Warn(msg string, args ...any)
		Error(msg string, args ...any)
	}
}

// NewMuxer creates a new FFmpeg muxer. If ffmpegPath is empty, "ffmpeg" is
// used (resolved via PATH). The ffprobe path is derived from the ffmpeg path.
func NewMuxer(ffmpegPath string, logger interface {
	Debug(msg string, args ...any)
	Info(msg string, args ...any)
	Warn(msg string, args ...any)
	Error(msg string, args ...any)
}) *Muxer {
	if ffmpegPath == "" {
		ffmpegPath = "ffmpeg"
	}
	return &Muxer{
		ffmpegPath:  ffmpegPath,
		ffprobePath: deriveFFprobePath(ffmpegPath),
		logger:      logger,
	}
}

// FFprobePath returns the resolved ffprobe binary path.
func (m *Muxer) FFprobePath() string {
	return m.ffprobePath
}

// deriveFFprobePath returns the ffprobe path corresponding to the given ffmpeg path.
func deriveFFprobePath(ffmpegPath string) string {
	if ffmpegPath == "" || ffmpegPath == "ffmpeg" {
		return "ffprobe"
	}
	dir := filepath.Dir(ffmpegPath)
	base := filepath.Base(ffmpegPath)
	// Case-insensitive match: Windows paths commonly carry arbitrary casing
	// ("FFmpeg.exe"), and a failed replacement would silently point the
	// ffprobe path at the ffmpeg binary itself.
	if idx := strings.Index(strings.ToLower(base), "ffmpeg"); idx >= 0 {
		return filepath.Join(dir, base[:idx]+"ffprobe"+base[idx+len("ffmpeg"):])
	}
	// Base name doesn't contain "ffmpeg" at all (renamed binary): fall back
	// to an ffprobe sibling in the same directory, preserving the extension.
	return filepath.Join(dir, "ffprobe"+filepath.Ext(base))
}

// MuxCopy muxes video and audio streams using codec copy (fast, no re-encoding).
func (m *Muxer) MuxCopy(ctx context.Context, videoPath, audioPath, outputPath string) error {
	return m.Mux(ctx, videoPath, audioPath, outputPath, nil)
}

// Mux performs FFmpeg muxing with optional trimming/encoding.
func (m *Muxer) Mux(ctx context.Context, videoPath, audioPath, outputPath string, opts *TrimOptions) error {
	// Validate inputs
	if videoPath != "" {
		if _, err := os.Stat(videoPath); err != nil {
			return fmt.Errorf("video file not found: %w", err)
		}
	}
	if audioPath != "" {
		if _, err := os.Stat(audioPath); err != nil {
			return fmt.Errorf("audio file not found: %w", err)
		}
	}

	// Ensure output directory exists
	if dir := filepath.Dir(outputPath); dir != "" {
		if err := os.MkdirAll(dir, 0o755); err != nil {
			return fmt.Errorf("create output directory: %w", err)
		}
	}

	// Determine strategy
	needsEncode := opts != nil && opts.UsePreciseTrim && hasTrim(opts) && (opts.CRF > 0 || opts.VideoBitrate > 0)
	needsTwoPass := needsEncode && opts.TwoPass && opts.VideoBitrate > 0

	if needsTwoPass {
		err := m.runTwoPassEncode(ctx, videoPath, audioPath, outputPath, opts)
		cleanupFailedMux(ctx, err, outputPath)
		return err
	}

	args := m.buildArgs(videoPath, audioPath, outputPath, opts, needsEncode)

	m.logger.Debug("ffmpeg", "args", strings.Join(args, " "))
	var err error
	if opts != nil && opts.ProgressFn != nil && hasTrim(opts) && opts.TrimDuration > 0 {
		err = m.runFFmpegWithProgress(ctx, args, opts.TrimDuration, opts.ProgressFn)
	} else {
		err = m.runFFmpeg(ctx, args)
	}
	cleanupFailedMux(ctx, err, outputPath)
	return err
}

// cleanupFailedMux removes the output file on mux failure, but preserves
// partial output when the failure is due to caller-initiated cancellation.
// Context cancel during mux is almost always a user shutdown — keeping the
// partial file lets the user decide whether to restart the mux or salvage
// what's there. True ffmpeg errors (invalid args, corrupt input, etc.)
// still get cleaned up so we don't leave broken MP4s on disk.
func cleanupFailedMux(ctx context.Context, err error, outputPath string) {
	if err == nil {
		return
	}
	if ctx.Err() != nil {
		// Parent context cancelled — preserve partial output.
		return
	}
	os.Remove(outputPath)
}

func (m *Muxer) buildArgs(videoPath, audioPath, outputPath string, opts *TrimOptions, needsEncode bool) []string {
	args := []string{"-y"} // Overwrite

	// Input args with optional trim
	if opts != nil && opts.TrimStartOffset > 0 {
		args = append(args, "-ss", fmt.Sprintf("%.3f", opts.TrimStartOffset))
	}
	if videoPath != "" {
		args = append(args, "-i", videoPath)
	}

	if audioPath != "" {
		if opts != nil && opts.TrimStartOffset > 0 {
			args = append(args, "-ss", fmt.Sprintf("%.3f", opts.TrimStartOffset))
		}
		args = append(args, "-i", audioPath)
	}

	// Duration limit
	if opts != nil && opts.TrimDuration > 0 {
		args = append(args, "-t", fmt.Sprintf("%.3f", opts.TrimDuration))
	}

	if needsEncode {
		args = m.appendEncodeArgs(args, audioPath, opts)
	} else {
		// Codec copy
		args = append(args, "-c", "copy")
	}

	// Output options
	args = append(args, "-movflags", "faststart")
	args = append(args, outputPath)

	return args
}

func (m *Muxer) appendEncodeArgs(args []string, audioPath string, opts *TrimOptions) []string {
	if opts.CRF > 0 {
		// CRF mode: re-encode video with constant quality
		args = append(args, "-c:v", "libx264", "-crf", strconv.Itoa(opts.CRF), "-preset", "slow")
		// Re-encode audio to keep sync with re-encoded video
		if opts.AudioBitrate > 0 {
			args = append(args, "-c:a", "aac", "-b:a", fmt.Sprintf("%dk", opts.AudioBitrate))
		} else {
			args = append(args, "-c:a", "aac")
		}
	} else {
		// ABR mode: target bitrate encoding
		if opts.VideoBitrate > 0 {
			args = append(args, "-c:v", "libx264", "-b:v", fmt.Sprintf("%dk", opts.VideoBitrate), "-preset", "fast")
		} else {
			args = append(args, "-c:v", "copy")
		}
		if audioPath != "" && opts.AudioBitrate > 0 {
			args = append(args, "-c:a", "aac", "-b:a", fmt.Sprintf("%dk", opts.AudioBitrate))
		} else if audioPath != "" {
			args = append(args, "-c:a", "copy")
		}
	}

	return args
}

func (m *Muxer) runFFmpeg(ctx context.Context, args []string) error {
	cmd := exec.CommandContext(ctx, m.ffmpegPath, args...)
	cmd.Stdout = nil

	// Capture stderr for error reporting with a 1MB cap.
	// If output exceeds the cap, only the last 512KB is kept.
	stderrBuf := &cappedBuffer{maxSize: 1 << 20, keepSize: 512 << 10}
	cmd.Stderr = stderrBuf

	if err := cmd.Run(); err != nil {
		if ctx.Err() != nil {
			return fmt.Errorf("ffmpeg cancelled: %w", ctx.Err())
		}
		stderr := stderrBuf.String()
		// Log last 500 chars of stderr for debugging
		if len(stderr) > 500 {
			stderr = stderr[len(stderr)-500:]
		}
		m.logger.Error("ffmpeg failed", "stderr", stderr)
		return fmt.Errorf("ffmpeg: %w (stderr: %s)", err, stderr)
	}

	return nil
}

func hasTrim(opts *TrimOptions) bool {
	return opts != nil && (opts.TrimStartOffset > 0 || opts.TrimDuration > 0)
}
